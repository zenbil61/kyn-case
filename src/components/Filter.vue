<template>
  <div class="filter-row">
    <label>{{ title }}</label>
    <div><slot></slot></div>
  </div>
</template>

<script>
export default {
  name: 'FilterComp',
  props: ['title'],
};
</script>

<style lang="scss">
.filter-row {
  border-radius: $radius;
  width: 100%;
  background: $gray;
  margin-bottom: 20px;
  padding: 20px;
  label {
    margin-bottom: 15px;
  }
}
</style>
