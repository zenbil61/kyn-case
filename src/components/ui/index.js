import Button from './Button.vue';
import Header from './Header.vue';
import HeaderItem from './HeaderItem.vue';
import NoResult from './NoResult.vue';
import Pagination from './Pagination.vue';
import PaginationItem from './PaginationItem.vue';
import Select from './Select.vue';

export { Button, Header, HeaderItem, NoResult, Pagination, PaginationItem, Select };
