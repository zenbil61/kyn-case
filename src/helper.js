function paginator(items, current_page, per_page_items) {
  let page = current_page || 1;
  let per_page = per_page_items || 10;
  let offset = (page - 1) * per_page;
  let paginatedItems = items.slice(offset).slice(0, per_page_items);
  return paginatedItems;
}
function wait(ms) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
const debounce = (func, wait) => {
  let timeout;

  return function executedFunction(...args) {
    const later = () => {
      timeout = null;

      func(...args);
    };

    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  };
};

export { paginator, debounce, wait };
