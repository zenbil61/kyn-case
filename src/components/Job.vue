<template>
  <div class="card job">
    <img class="card-img-top" />
    <div class="card-body">
      <h5 class="card-title">{{ props.title }}</h5>
      <p class="card-text">
        {{ props.description }}
      </p>
      <Button type="primary" @click="emit('click')">Detay</Button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'FilterComp',
};
</script>
<script setup>
import { Button } from '../components/ui';
import { defineEmits, defineProps } from 'vue';

const emit = defineEmits(['click']);
const props = defineProps(['title', 'description']);
</script>

<style lang="scss">
.card.job {
  margin-top: 30px;
  border-radius: $radius;
}
</style>
