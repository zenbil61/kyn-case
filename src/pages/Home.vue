<template>
  <div class="landing">
    <div class="landing-detail">
      <h1>Aradığın işe anında kavuş</h1>
      <p>Öne çıkan ilanları görmek için hemen şehrine göre ara ve ilanları gör</p>
    </div>
    <div class="row">
      <div class="col-md-4">
        <Select v-model="selectedCity" :data="store.state.cities" />
      </div>
      <div class="col">
        <input v-model="searchText" type="text" class="form-control search-input" placeholder="Lütfen Arama Yapın.." />
      </div>
      <div class="col-md-2">
        <Button @click="handleSearch">Ara</Button>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'HomeComp',
};
</script>
<script setup>
import { ref, onMounted } from 'vue';
import { useRouter } from 'vue-router';
import { ROUTE, ACTIONS } from '../constant';
import { Button, Select } from '../components/ui';
import { useStore } from 'vuex';

const router = useRouter();
const store = useStore();
const searchText = ref(null);
const selectedCity = ref(null);

const handleSearch = () => {
  const query = {};

  if (searchText.value) query.search = searchText.value;
  if (selectedCity.value) query.city = selectedCity.value;

  router.push({
    path: ROUTE.PAGE_JOBS,
    query,
  });
};

onMounted(() => {
  store.dispatch(ACTIONS.FETCH_CITIES);
});
</script>

<style lang="scss">
.landing {
  background: $gray;
  padding: 50px;
  &-detail {
    p {
      margin: 50px 0px;
    }
  }
}
</style>
