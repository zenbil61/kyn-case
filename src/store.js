import { createStore } from 'vuex';
import { ACTIONS, MUTATIONS } from './constant';
import axios from 'axios';
import { wait } from './helper';
const BASE_API_URL = process.env.VUE_APP_BASE_API_URL;

const store = createStore({
  state() {
    return {
      jobs_data: {},
      jobs_data_loading: true,
      cities: [],
      job: {},
      job_loading: true,
    };
  },
  mutations: {
    [MUTATIONS.SET_JOBS]: (state, data) => (state.jobs_data = data),
    [MUTATIONS.SET_JOBS_LOADING]: (state, data) => (state.jobs_data_loading = data),
    [MUTATIONS.SET_JOB]: (state, data) => (state.job = data),
    [MUTATIONS.SET_JOB_LOADING]: (state, data) => (state.job_loading = data),
    [MUTATIONS.SET_CITIES]: (state, data) => (state.cities = data),
  },
  actions: {
    async [ACTIONS.FETCH_JOBS]({ commit }, request) {
      commit(MUTATIONS.SET_JOBS_LOADING, true);
      await wait(500);
      const { data } = await axios.get(`${BASE_API_URL}/jobs`, {
        params: request,
      });
      commit(MUTATIONS.SET_JOBS, data.data);
      commit(MUTATIONS.SET_JOBS_LOADING, false);
    },
    async [ACTIONS.FETCH_CITIES]({ commit }) {
      const { data } = await axios.get(`${BASE_API_URL}/cities`);
      commit(MUTATIONS.SET_CITIES, data.data);
    },
    async [ACTIONS.FETCH_JOB]({ commit }, { id }) {
      commit(MUTATIONS.SET_JOB_LOADING, true);
      await wait(500);

      const { data } = await axios.get(`${BASE_API_URL}/jobs/detail`, {
        params: { id },
      });
      commit(MUTATIONS.SET_JOB, data.data);
      commit(MUTATIONS.SET_JOB_LOADING, false);
    },
  },
  getters: {
    cityName: (state) => (id) => {
      return state.cities.find((city) => city.value == id).text;
    },
  },
});

export default store;
