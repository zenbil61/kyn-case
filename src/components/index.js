import Filter from './Filter.vue';
import Job from './Job.vue';

export { Filter, Job };
