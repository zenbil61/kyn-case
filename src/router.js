import { createRouter, createWebHistory } from 'vue-router';

import Home from './pages/Home.vue';
import Detail from './pages/Detail.vue';
import List from './pages/List.vue';

import { ROUTE } from './constant';
const routes = [
  { path: '/', component: Home },
  { path: ROUTE.PAGE_JOBS, component: List },
  {
    path: ROUTE.PAGE_JOBS_DETAIL,
    component: Detail,
  },
];
const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;
