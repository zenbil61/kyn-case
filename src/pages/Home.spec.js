import { mount } from '@vue/test-utils';
import Home from './Home.vue';
import { Button } from '@/components/ui';
import router from '@/router';
import store from '@/store';
import { wait } from '@/helper';

it('renders correctly', async () => {
  const wrapper = mount(Home, {
    global: {
      plugins: [router, store],
    },
  });
  await wait(2000);
  expect(wrapper.element).toMatchSnapshot();
  wrapper.findComponent(Button).trigger('click');
});
