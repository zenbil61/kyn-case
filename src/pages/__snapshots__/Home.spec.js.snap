// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="landing"
>
  <div
    class="landing-detail"
  >
    <h1>
      Aradığın işe anında kavuş
    </h1>
    <p>
      Öne çıkan ilanları görmek için hemen şehrine göre ara ve ilanları gör
    </p>
  </div>
  <div
    class="row"
  >
    <div
      class="col-md-4"
    >
      <select
        class="form-select"
      >
        <option
          disabled=""
          value=""
        >
          Seçim Yapın
        </option>
        
        <option
          value="34"
        >
          İstanbul
        </option>
        <option
          value="6"
        >
          Ankara
        </option>
        <option
          value="35"
        >
          İzmir
        </option>
        
      </select>
    </div>
    <div
      class="col"
    >
      <input
        class="form-control search-input"
        placeholder="Lütfen Arama Yapın.."
        type="text"
      />
    </div>
    <div
      class="col-md-2"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        
        Ara
        
      </button>
    </div>
  </div>
</div>
`;
